proof.dfy

```diff
@@ -148,6 +211,138 @@ module Proof {
+    var recordedPreparesSenders := h_v.workingWindow.preparesRcvd[new_msg.seqID].Keys;
+    assert |recordedPreparesSenders| >= c.clusterConfig.AgreementQuorum();
+    var prepares := sentPreparesForSeqID(c, v, new_msg.seqID, new_msg.clientOp);
+    assert recordedPreparesSenders <= sendersOf(prepares) by {
+      forall sender | sender in recordedPreparesSenders ensures sender in sendersOf(prepares) {
+        var msg := h_v.workingWindow.preparesRcvd[new_msg.seqID][sender];
+        assert msg.sender == sender;
+        assert msg.seqID == new_msg.seqID;
+        assert msg.clientOp == new_msg.clientOp;
+        assert msg in v.network.sentMsgs;
+        assert msg in prepares;
+      }
+    }
+
+    var prepares' := sentPreparesForSeqID(c, v, new_msg.seqID, new_msg.clientOp);
+    Library.SubsetCardinality(prepares, prepares');
+
+    assert forall sender | sender in h_v.workingWindow.preparesRcvd[new_msg.seqID]
+                         :: h_v.workingWindow.preparesRcvd[new_msg.seqID][sender].clientOp == new_msg.clientOp;
+
+    assert h_v.workingWindow.prePreparesRcvd[new_msg.seqID].value.clientOp == new_msg.clientOp;
+
+    assert Replica.QuorumOfPrepares(h_c, h_v, old_msg.seqID);
+
+    assert h_v.workingWindow.prePreparesRcvd[old_msg.seqID].value.clientOp == old_msg.clientOp;
+
+    assert forall sender | sender in h_v.workingWindow.preparesRcvd[old_msg.seqID]
+                         :: h_v.workingWindow.preparesRcvd[old_msg.seqID][sender].clientOp == old_msg.clientOp;
+
+    Library.SubsetCardinality(recordedPreparesSenders, sendersOf(prepares));
+    assert |sendersOf(prepares)| >= c.clusterConfig.AgreementQuorum();
+    SendersAreAlwaysFewerThanMessages(prepares);
+    assert |prepares| >= |sendersOf(prepares)|;
+    assert |prepares| >= c.clusterConfig.AgreementQuorum();
+    assert QuorumOfPreparesInNetwork(c, v', new_msg.seqID, new_msg.clientOp);
+
+    assert old_msg.seqID == new_msg.seqID;
+    assert old_msg.Commit? && new_msg.Commit?;
+    assert old_msg.sender == new_msg.sender;
+    assert old_msg.view == new_msg.view;
+    assert old_msg.clientOp == new_msg.clientOp;
+
+    assert old_msg == new_msg;
+  }
+
+  lemma HonestReplicasLockOnCommitForGivenViewLemma(c: Constants, v:Variables, v':Variables, 
+                                                    step:Step, h_step:Replica.Step)
+    requires Inv(c, v)
+    requires NextStep(c, v, v', step)
+    requires c.IsReplica(step.id)
+    requires  var h_c := c.hosts[step.id].replicaConstants;
+              var h_v := v.hosts[step.id].replicaVariables;
+              var h_v' := v'.hosts[step.id].replicaVariables;
+              && Replica.NextStep(h_c, h_v, h_v', step.msgOps, h_step)
+              && h_step.SendCommitStep?
+              && Replica.SendCommit(h_c, h_v, h_v', step.msgOps, h_step.seqID)
+    ensures HonestReplicasLockOnCommitForGivenView(c, v')
+  {
+    forall msg1, msg2 | 
+      && msg1 in v'.network.sentMsgs 
+      && msg2 in v'.network.sentMsgs 
```

Comment:
Is this correct? Both messages are sourced from the new state. Also, later on all possibilities are examined - (msg1 and msg2 exist in either new/old state). Doesn't that imply that there's nondeterminism about where msg1/msg2 exist so these checks should be removed?

Reply:
I guess it's correct, msg1/msg2 should always exist in the new state.

Reply:
Good point, we don't delete messages from the network, this set only grows. Therefore if both msg1 and msg2 are in v'.network.sentMsgs, it could be because both were in v.network.sentMsgs, or one of them was there, and the other was sent by the step generating v' from v. Basically for the sets we have v.network.sentMsgs <= v'.network.sentMsgs.